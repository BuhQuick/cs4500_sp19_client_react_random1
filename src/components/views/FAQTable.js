import React from 'react'

const FAQTable = (props) => {
  return (
    <div>
      <h3>Frequently Asked Questions</h3>
      <table className="table">
        <tr>
            <th>Title</th>
            <th>Question</th>
            <th></th>
        </tr>
        <tr>
            <th>
                <input
                    type="text"
                    id="titleInput"
                    placeholder="Title"
                    title="Type in a name">
                </input>
            </th>
            <th>
                <input
                    type="text"
                    id="questionInput"
                    placeholder="Question"
                    title="Type in a name">
                </input>
            </th>
            <th>
                <button onClick={() => props.addFn()}>Add</button>
                <button>Save</button>
            </th>
        </tr>
        <tbody>
          {
           props.faqs.map(faq =>
             <tr key={faq.id}>
               <td>{faq.title}</td>
               <td>{faq.question}</td>
               <td>
                    <button onClick={() => props.deleteFn(faq.id)}>Delete</button>
                   <button onClick={() => props.updateFn(faq.id)}>Edit</button>
              </td>
            </tr>
          )
          }
        </tbody>
      </table>
    </div>
  )
}

export default FAQTable;
