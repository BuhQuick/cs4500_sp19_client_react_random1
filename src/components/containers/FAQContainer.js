import React from 'react'
import FAQService from '../../services/FAQService'
import FAQSearch from '../views/FAQSearch'
import FAQTable from '../views/FAQTable'
import FAQPagination from '../views/FAQPagination'

class FAQContainer extends React.Component {
    constructor(props) {
        super(props)
        this.faqService = FAQService.getInstance()
        this.state = {
            faqs: [],
            searchButtonDisabled: false,
            pageNumber: 1,
            itemsPerPage: 10
        }
        // bind functions here
        this.refreshFAQs = this.refreshFAQs.bind(this)
        this.searchFAQs = this.searchFAQs.bind(this)
        this.clearSearch = this.clearSearch.bind(this)
        this.deleteFAQ = this.deleteFAQ.bind(this)
        this.nextPage = this.nextPage.bind(this)
        this.previousPage = this.previousPage.bind(this)
        this.onChangeItemsPerPage = this.onChangeItemsPerPage.bind(this)
        this.createPageButtons = this.createPageButtons.bind(this)
    }

    componentDidMount() {
        this.faqService
            .findAllFAQs()
            .then(results =>
                this.setState({
                    faqs: results.slice(),
                    searchButtonDisabled: false
                })
            )
    }

    refreshFAQs() {
        this.faqService.findAllFAQs().then(results =>
            this.setState({
                faqs: results
            }))
    }

    deleteFAQ(id) {
        console.log('Delete FAQ Click happened');
        this.faqService.deleteFAQ(id)
            .then(() => this.refreshFAQs());
    }


    searchFAQs() {
        const titleInput = document.getElementById('titleInput');
        const questionInput = document.getElementById('questionInput');
        var title = titleInput.value
        var question = questionInput.value
        debugger;
        if(title === "" && question === "") {
            this.faqService
                .findAllFAQs()
                .then(results =>
                    this.setState(prevState => ({
                        faqs: results,
                        searchButtonDisabled: false
                    }))
                )
            return;
        }
        this.faqService
            .filterFAQs({title: title, question: question})
            .then(results =>
                this.setState(prevState => ({
                    faqs: results,
                    searchButtonDisabled: !prevState.searchButtonDisabled
                }))
            )
    }
    clearSearch() {
        const titleInput = document.getElementById('titleInput');
        const questionInput = document.getElementById('questionInput');
        titleInput.value = ""
        questionInput.value = ""
        this.faqService
            .findAllFAQs()
            .then(results =>
                this.setState(prevState => ({
                    faqs: results,
                    searchButtonDisabled: false
                }))
            )
    }

    onChangeItemsPerPage() {
        var currButtons = document.getElementById("pageButtons").innerHTML = "";
        var eID = document.getElementById("selector");
        var pageText = eID.options[eID.selectedIndex].text;

        if (pageText == "10") {
            this.setState({itemsPerPage: 10, pageNumber: 1})
        }
        if (pageText == "25") {
            this.setState({itemsPerPage: 25, pageNumber: 1})
        }
        if (pageText == "50") {
            this.setState({itemsPerPage: 50, pageNumber: 1})
        }
        if (pageText == "All") {
            this.setState({itemsPerPage: 1000000, pageNumber: 1})
        }
    }
    previousPage() {
        if (this.state.pageNumber != 1) {
            var temp = this.state.pageNumber - 1;
            this.setState({pageNumber: temp })
        }
    }
    nextPage() {
        var totalPages = Math.ceil(this.state.faqs.length / this.state.itemsPerPage);
        if (this.state.pageNumber + 1 <= totalPages) {
            var temp = this.state.pageNumber + 1;
            this.setState({pageNumber: temp })
        }
    }

    createPageButtons() {
        var totalPages = Math.ceil(this.state.faqs.length / this.state.itemsPerPage);
        var i;
        for (i = 1; i < totalPages + 1; i++) {
            if ((document.getElementById("Button " + i)) == null)
            {
                console.log(i);
                var button = document.createElement("BUTTON");
                var buttonText = document.createTextNode("" + i);
                button.appendChild(buttonText);
                button.type = "button";
                button.id = "Button " + i;
                button.value = i;
                button.onclick = function(e) {
                    this.setState({pageNumber: e.target.value})
                }.bind(this);
                document.getElementById("pageButtons").appendChild(button);
            }
        }
    }

    render() {
        let result = (
            <div>
                <FAQTable
                    faqs={this.state.faqs}
                    deleteFn={this.deleteFAQ}
                    pageNumber={this.state.pageNumber}
                    itemsPerPage={this.state.itemsPerPage}/>
                <FAQSearch
                    disabled={this.state.searchButtonDisabled}
                    searchFn={this.searchFAQs}
                    clearFn={this.clearSearch}
                />
                <FAQPagination id="Pagination"
                    previousFn={this.previousPage}
                    nextFn={this.nextPage}
                    changeItemsPerPageFn={this.onChangeItemsPerPage}
                    createButtonsFn={this.createPageButtons}/>
            </div>
        )
        return result;
    }
}

export default FAQContainer;
