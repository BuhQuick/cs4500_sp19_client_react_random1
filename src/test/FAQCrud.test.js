import React from 'react';
import FAQTable from '../components/views/FAQTable';
import FAQContainer from '../components/containers/FAQContainer';
import renderer from 'react-test-renderer';
import faqData from '../mock/data/faq.mock.json';



it('Render FAQTable page 1, 10 items/page', () => {
    const tree = renderer.create(
        <FAQTable
            faqs={faqData}
            addFn={FAQContainer.addFAQ}
            updateFn={FAQContainer.updateFAQ}
            deleteFn={FAQContainer.deleteFAQ}
            pageNumber={1}
            itemsPerPage={10}/>
    ).toJSON();
    expect(tree).toMatchSnapshot();
});

it('Render FAQTable page 2, 5 items/page', () => {
    const tree = renderer.create(
        <FAQTable
            faqs={faqData}
            addFn={FAQContainer.addFAQ}
            updateFn={FAQContainer.updateFAQ}
            deleteFn={FAQContainer.deleteFAQ}
            pageNumber={2}
            itemsPerPage={5}/>
    ).toJSON();
    expect(tree).toMatchSnapshot();
});
