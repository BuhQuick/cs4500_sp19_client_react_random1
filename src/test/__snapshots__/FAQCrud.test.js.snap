// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render FAQTable page 1, 10 items/page 1`] = `
<div>
  <h3>
    Frequently Asked Questions
  </h3>
  <table
    className="table"
  >
    <tr>
      <th>
        Title
      </th>
      <th>
        Question
      </th>
      <th />
    </tr>
    <tr>
      <th>
        <input
          id="titleInput"
          placeholder="Title"
          title="Type in a name"
          type="text"
        />
      </th>
      <th>
        <input
          id="questionInput"
          placeholder="Question"
          title="Type in a name"
          type="text"
        />
      </th>
      <th>
        <button
          onClick={[Function]}
        >
          Add
        </button>
        <button>
          Save
        </button>
      </th>
    </tr>
    <tbody>
      <tr>
        <td>
          Q1
        </td>
        <td>
          What is my first question?
        </td>
        <td>
          <button
            onClick={[Function]}
          >
            Delete
          </button>
          <button
            onClick={[Function]}
          >
            Edit
          </button>
        </td>
      </tr>
      <tr>
        <td>
          Q2
        </td>
        <td>
          What is my second question?
        </td>
        <td>
          <button
            onClick={[Function]}
          >
            Delete
          </button>
          <button
            onClick={[Function]}
          >
            Edit
          </button>
        </td>
      </tr>
      <tr>
        <td>
          Q3
        </td>
        <td>
          What is my third question?
        </td>
        <td>
          <button
            onClick={[Function]}
          >
            Delete
          </button>
          <button
            onClick={[Function]}
          >
            Edit
          </button>
        </td>
      </tr>
      <tr>
        <td>
          Q4
        </td>
        <td>
          What is my fourth question?
        </td>
        <td>
          <button
            onClick={[Function]}
          >
            Delete
          </button>
          <button
            onClick={[Function]}
          >
            Edit
          </button>
        </td>
      </tr>
      <tr>
        <td>
          Q5
        </td>
        <td>
          What is my fifth question?
        </td>
        <td>
          <button
            onClick={[Function]}
          >
            Delete
          </button>
          <button
            onClick={[Function]}
          >
            Edit
          </button>
        </td>
      </tr>
      <tr>
        <td>
          Q6
        </td>
        <td>
          What is my sixth question?
        </td>
        <td>
          <button
            onClick={[Function]}
          >
            Delete
          </button>
          <button
            onClick={[Function]}
          >
            Edit
          </button>
        </td>
      </tr>
      <tr>
        <td>
          Q7
        </td>
        <td>
          What is my seventh question?
        </td>
        <td>
          <button
            onClick={[Function]}
          >
            Delete
          </button>
          <button
            onClick={[Function]}
          >
            Edit
          </button>
        </td>
      </tr>
      <tr>
        <td>
          Q8
        </td>
        <td>
          What is my eight question?
        </td>
        <td>
          <button
            onClick={[Function]}
          >
            Delete
          </button>
          <button
            onClick={[Function]}
          >
            Edit
          </button>
        </td>
      </tr>
      <tr>
        <td>
          Q9
        </td>
        <td>
          What is my ninth question?
        </td>
        <td>
          <button
            onClick={[Function]}
          >
            Delete
          </button>
          <button
            onClick={[Function]}
          >
            Edit
          </button>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`Render FAQTable page 2, 5 items/page 1`] = `
<div>
  <h3>
    Frequently Asked Questions
  </h3>
  <table
    className="table"
  >
    <tr>
      <th>
        Title
      </th>
      <th>
        Question
      </th>
      <th />
    </tr>
    <tr>
      <th>
        <input
          id="titleInput"
          placeholder="Title"
          title="Type in a name"
          type="text"
        />
      </th>
      <th>
        <input
          id="questionInput"
          placeholder="Question"
          title="Type in a name"
          type="text"
        />
      </th>
      <th>
        <button
          onClick={[Function]}
        >
          Add
        </button>
        <button>
          Save
        </button>
      </th>
    </tr>
    <tbody>
      <tr>
        <td>
          Q6
        </td>
        <td>
          What is my sixth question?
        </td>
        <td>
          <button
            onClick={[Function]}
          >
            Delete
          </button>
          <button
            onClick={[Function]}
          >
            Edit
          </button>
        </td>
      </tr>
      <tr>
        <td>
          Q7
        </td>
        <td>
          What is my seventh question?
        </td>
        <td>
          <button
            onClick={[Function]}
          >
            Delete
          </button>
          <button
            onClick={[Function]}
          >
            Edit
          </button>
        </td>
      </tr>
      <tr>
        <td>
          Q8
        </td>
        <td>
          What is my eight question?
        </td>
        <td>
          <button
            onClick={[Function]}
          >
            Delete
          </button>
          <button
            onClick={[Function]}
          >
            Edit
          </button>
        </td>
      </tr>
      <tr>
        <td>
          Q9
        </td>
        <td>
          What is my ninth question?
        </td>
        <td>
          <button
            onClick={[Function]}
          >
            Delete
          </button>
          <button
            onClick={[Function]}
          >
            Edit
          </button>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
